src/bin/devmgr/src/nvme_test.rs

```diff
@@ -57,8 +78,273 @@ impl RequestDriver for NvmeQueue {
     const NUM_IDS: usize = 8;
 }
 
-async fn test<'a>(ctrl: Arc<NvmeController>) {
-    println!("starting req test");
+async fn test3<'a>(ctrl: Arc<NvmeController>) {
+    let int = ctrl.device_ctrl.allocate_interrupt().unwrap();
+}
+
+async fn test2<'a>(ctrl: Arc<NvmeController>) {
+    loop {
+        let (mp, msg) = ctrl.device_ctrl.next_msg(MailboxPriority::Idle).await;
+        println!("mailbox message: {:?} {}", mp, msg);
+    }
+}
+
+struct NvmeDmaRegion<'a, T: DeviceSync>(DmaRegion<'a, T>);
+
+impl<'a, T: DeviceSync> PhysicalPageCollection for NvmeDmaRegion<'a, T> {
+    fn get_prp_list_or_buffer(&mut self) -> Option<nvme::ds::cmd::PrpListOrBuffer> {
+        todo!()
+    }
+
+    fn get_dptr(&mut self, _sgl_allowed: bool) -> Option<nvme::ds::queue::subentry::Dptr> {
+        let pin = self.0.pin().unwrap();
+        Some(nvme::ds::queue::subentry::Dptr::Prp(
+            pin[0].addr().into(),
+            0,
+        ))
+    }
+}
+
+struct NvmeRequester {
+    subq: Mutex<SubmissionQueue>,
+    comq: Mutex<CompletionQueue>,
+    sub_bell: *const VolatileCell<u32>,
+    com_bell: *const VolatileCell<u32>,
+}
+
+unsafe impl Send for NvmeRequester {}
+unsafe impl Sync for NvmeRequester {}
+
+impl NvmeRequester {
+    fn check_completions(&self) -> Vec<ResponseInfo<CommonCompletion>> {
+        let mut comq = self.comq.lock().unwrap();
+        let mut resps = Vec::new();
+        let mut new_head = None;
+        let mut new_bell = None;
+        while let Some((bell, resp)) = comq.get_completion::<CommonCompletion>() {
+            let id: u16 = resp.command_id().into();
+            //println!("got completion for {} {} {}", resp.new_sq_head(), bell, id);
+            resps.push(ResponseInfo::new(resp, id as u64, false));
+            new_head = Some(resp.new_sq_head());
+            new_bell = Some(bell);
+        }
+
+        if let Some(head) = new_head {
+            self.subq.lock().unwrap().update_head(head);
+        }
+
+        if let Some(bell) = new_bell {
+            unsafe { self.com_bell.as_ref().unwrap().set(bell as u32) }
+        }
+
+        resps
+    }
+}
+
+#[async_trait::async_trait]
+impl RequestDriver for NvmeRequester {
+    type Request = CommonCommand;
+
+    type Response = CommonCompletion;
+
+    type SubmitError = ();
+
+    async fn submit(
+        &self,
+        reqs: &mut [SubmitRequest<Self::Request>],
+    ) -> Result<(), Self::SubmitError> {
+        let mut sq = self.subq.lock().unwrap();
+        let mut tail = None;
+        for sr in reqs.iter_mut() {
+            let cid = (sr.id() as u16).into();
+            sr.data_mut().set_cid(cid);
+            tail = sq.submit(sr.data());
+            assert!(tail.is_some());
+        }
+        if let Some(tail) = tail {
+            unsafe {
+                self.sub_bell.as_ref().unwrap().set(tail as u32);
+            }
+        }
+        Ok(())
+    }
+
+    fn flush(&self) {}
+
+    const NUM_IDS: usize = 32;
```

Comment:
Is this a leftover of something?

Reply:
The flush function? It's not needed for this implementation of the RequesterDriver trait (since submit flushes), but it still has to be implemented to satisfy the compiler.